server-go/filtering/parser/parser_test.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2018 The Grafeas Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// package parser provides methods to parse filter sources to CEL-based ASTs.
+package parser
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"testing"
+
+	pb "github.com/golang/protobuf/proto"
+	expr "github.com/google/cel-spec/proto/v1"
+	"github.com/grafeas/grafeas/server-go/filtering/common"
+)
+
+const (
+	DiagnosticsDelimiter = "\nDiagnostics:\n"
+	NoResult             = "<no result>"
+	InputOutputDelimiter = "\n==================================================\n"
+	TestCaseDelimiter    = "\n\n"
+)
+
+type baseline struct {
+	expected *expr.ParsedExpr
+	errors   string
+	source   common.Source
+}
+
+func TestParse_Complex(t *testing.T) {
```

Comment:
This set of nested functions seems relatively complicated - is there any way we could just have a single test case which loads all the files and runs the tests inside a series of nested for loops?  (That way you can read the testcase without having to go look at function definitions).

Reply:
If you need a message diffing library https://github.com/d4l3k/messagediff works.

re: IDE's and rerunning tests - I don't care strongly about whether there is a single table of tests, or you load each file in a separate test, I just just strongly encourage having tests be as simple as possible with minimal usage of helper functions and layers of abstraction.

The idea is that you want tests to be 'obviously correct' on inspection. Layers of function calls make it impossible to have this, and instead you have to go read each function and it's easier for bugs to slip in.

However I also approved this, so feel free to ignore my suggestion - this is mostly minor test style and I'm happy the tests exist at all :D

Reply:
Thanks, I'm not sure I have time to restructure the tests just now, but will keep this in mind for the future.